modules/tensor_mechanics/src/materials/DiscreteRRIsotropicPlasticity.C

```diff
@@ -0,0 +1,150 @@
+  params.addParam<Real>("hardening_constant", "Hardening slope");
+  params.addParam<FunctionName>("hardening_function", "True stress as a function of plastic strain");
+  params.addCoupledVar("temp", "Coupled Temperature");
+
+  return params;
+}
+
+
+DiscreteRRIsotropicPlasticity::DiscreteRRIsotropicPlasticity( const InputParameters & parameters)
+  :DiscreteRadialReturnStressIncrement(parameters),
+   _yield_stress_function(isParamValid("yield_stress_function") ? &getFunction("yield_stress_function") : NULL),
+   _yield_stress(isParamValid("yield_stress") ? getParam<Real>("yield_stress") : 0),
+   _hardening_constant(isParamValid("hardening_constant") ? getParam<Real>("hardening_constant") : 0),
+   _hardening_function(isParamValid("hardening_function") ? dynamic_cast<PiecewiseLinear*>(&getFunction("hardening_function")) : NULL),
+
+   _plastic_strain(declareProperty<RankTwoTensor>("plastic_strain")),
+   _plastic_strain_old(declarePropertyOld<RankTwoTensor>("plastic_strain")),
+   _scalar_plastic_strain(declareProperty<Real>("scalar_plastic_strain")),
+   _scalar_plastic_strain_old(declarePropertyOld<Real>("scalar_plastic_strain")),
+
+   _hardening_variable(declareProperty<Real>("hardening_variable")),
+   _hardening_variable_old(declarePropertyOld<Real>("hardening_variable")),
+   _shear_modulus(declareProperty<Real>("shear_modulus")),
+   _temperature(isCoupled("temp") ? coupledValue("temp") : _zero)
+{
+  if (isParamValid("yield_stress") && _yield_stress <= 0)
+    mooseError("Yield stress must be greater than zero");
+
+  if ( _yield_stress_function == NULL && !isParamValid("yield_stress"))
+    mooseError("Either yield_stress or yield_stress_function must be given");
+
+  if (!isParamValid("hardening_constant") && !isParamValid("hardening_function"))
+    mooseError("Either hardening_constant or hardening_function must be defined");
+
+  if (isParamValid("hardening_constant") && isParamValid("hardening_function"))
+    mooseError("Only the hardening_constant or only the hardening_function can be defined but not both");
+
+  if (isParamValid("hardening_function") && !_hardening_function)
+    mooseError("The hardening_function must be PiecewiseLinear");
+}
+
+void
+DiscreteRRIsotropicPlasticity::resetQpProperties()
+{
+  _hardening_variable[_qp] = _hardening_variable_old[_qp];
+  _plastic_strain[_qp] = _plastic_strain_old[_qp];
+  _scalar_plastic_strain[_qp] = 0;
+  if (!_hardening_function)
+    _scalar_plastic_strain_old[_qp] = 0;
+
+  DiscreteRadialReturnStressIncrement::resetQpProperties();
+}
+
+void
+DiscreteRRIsotropicPlasticity::computeStressInitialize(Real effectiveTrialStress)
+{
+  getIsotropicShearModulus();
+  computeYieldStress();
+  _yield_condition = effectiveTrialStress - _hardening_variable_old[_qp] - _yield_stress;
+}
+
+Real
+DiscreteRRIsotropicPlasticity::computeResidual(Real effectiveTrialStress, Real scalar)
+{
+  Real residual = 0;
+  _hardening_slope = 0;
```

Comment:
I **strongly** suggest to write all `Real` number literals _with_ a decimal point! Subtle bugs can arise from using integer literals in place of floating point literals. I just fixed one two weeks ago where the compiler performed an integer division instead of a float division. It is just better style and more expressive to use the proper literals!